pkg/cmd/roachtest/tests/mixed_version_change_replicas.go

```diff
@@ -0,0 +1,312 @@
+			require.NoError(t, row.Scan(&count))
+			require.Equal(t, expectRows, count)
+		}
+	}
+
+	// scatterTableStep scatters the replicas and leases for a table.
+	scatterTableStep := func(table string) versionStep {
+		return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+			gateway := randomNodeID()
+			t.Status(fmt.Sprintf("scattering table %s via n%d", table, gateway))
+			conn := u.c.Conn(ctx, t.L(), gateway)
+			defer conn.Close()
+			_, err = conn.ExecContext(ctx, `ALTER TABLE test SCATTER`)
+			require.NoError(t, err)
+		}
+	}
+
+	// changeReplicasRelocateFromNodeStep moves all table replicas from the given
+	// node onto any other node that doesn't already have a replica, using ALTER
+	// TABLE RELOCATE via a random gateway node.
+	changeReplicasRelocateFromNodeStep := func(table string, nodeID int) versionStep {
+		return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+
+			// Disable the replicate queue, but re-enable it when we're done.
+			setReplicateQueueEnabled := func(enabled bool) {
+				for n := 1; n <= nodeCount; n++ {
+					conn := u.c.Conn(ctx, t.L(), n)
+					defer conn.Close()
+					_, err := conn.ExecContext(ctx,
+						`SELECT crdb_internal.kv_set_queue_active('replicate', $1)`, enabled)
+					require.NoError(t, err)
+				}
+			}
+			setReplicateQueueEnabled(false)
+			defer setReplicateQueueEnabled(true)
+
+			// Relocate replicas to other nodes which don't have one, in random order.
+			targets := []int{}
+			for n := 1; n <= nodeCount; n++ {
+				if n != nodeID {
+					targets = append(targets, n)
+				}
+			}
+			rand.Shuffle(len(targets), func(i, j int) {
+				targets[i], targets[j] = targets[j], targets[i]
+			})
+
+			for _, target := range targets {
+				gateway := randomNodeID()
+				conn := u.c.Conn(ctx, t.L(), gateway)
+				defer conn.Close()
+
+				var rangeErrors map[int]string
+				for attempt := 1; attempt <= 5; attempt++ {
+					t.Status(fmt.Sprintf(
+						"moving replicas from n%d to n%d via gateway n%d using ALTER TABLE RELOCATE (attempt %d)",
+						nodeID, target, gateway, attempt))
+
+					var rangeID int
+					var pretty, result string
+					rows, err := conn.QueryContext(ctx, `ALTER RANGE RELOCATE FROM $1::int TO $2::int FOR `+
+						`SELECT range_id FROM [SHOW RANGES FROM TABLE test] `+
+						`WHERE $1::int = ANY(replicas) AND $2::int != ALL(replicas)`, nodeID, target)
+					require.NoError(t, err)
+
+					rangeErrors = map[int]string{}
+					for rows.Next() {
+						require.NoError(t, rows.Scan(&rangeID, &pretty, &result))
+						if result != "ok" {
+							rangeErrors[rangeID] = result
+						}
+					}
+					require.NoError(t, rows.Err())
+					if len(rangeErrors) == 0 {
+						break
+					}
+					t.Status(fmt.Sprintf("%d ranges failed", len(rangeErrors)))
+				}
+
+				if len(rangeErrors) > 0 {
+					for rangeID, result := range rangeErrors {
+						t.Status(fmt.Sprintf("failed to move r%d from n%d to n%d via n%d: %s",
```

Comment:
Why `t.Status`? Don't you want `t.L().Printf(...)`?

Reply:
Just cargo culting, updated.